Userland/Libraries/LibWeb/Layout/FlexFormattingContext.cpp

```diff
@@ -19,49 +24,666 @@ FlexFormattingContext::~FlexFormattingContext()
+            // E
+            // FIXME: This is probably too naive.
+            if (has_definite_main_size(child_box)) {
+                flex_item.flex_base_size = specified_main_size(child_box);
+            } else {
+                layout_for_maximum_main_size(child_box);
+                flex_item.flex_base_size = calculated_main_size(child_box);
+            }
+        }
+
+        auto clamp_min = has_main_min_size(child_box)
+            ? specified_main_min_size(child_box)
+            : 0;
+        auto clamp_max = has_main_max_size(child_box)
+            ? specified_main_max_size(child_box)
+            : NumericLimits<float>::max();
+
+        flex_item.hypothetical_main_size = clamp(flex_item.flex_base_size, clamp_min, clamp_max);
+    }
+
+    /* 4. Determine the main size of the flex container */
+    if (!main_is_constrained || main_available_size == 0) {
+        /* Uses https://www.w3.org/TR/css-flexbox-1/#intrinsic-main-sizes */
+        // 9.9.1
+        // 1.
+        float largest_max_content_flex_fraction = 0;
+        for (auto& flex_item : flex_items) {
+            // FIXME: This needs some serious work.
+            float max_content_contribution = calculated_main_size(flex_item.box);
+            float max_content_flex_fraction = max_content_contribution - flex_item.flex_base_size;
+            if (max_content_flex_fraction > 0) {
+                max_content_flex_fraction /= max(flex_item.box.computed_values().flex_grow_factor().value_or(1), 1.0f);
+            } else {
+                max_content_flex_fraction /= max(flex_item.box.computed_values().flex_shrink_factor().value_or(1), 1.0f) * flex_item.flex_base_size;
+            }
+            flex_item.max_content_flex_fraction = max_content_flex_fraction;
+
+            if (max_content_flex_fraction > largest_max_content_flex_fraction)
+                largest_max_content_flex_fraction = max_content_flex_fraction;
+        }
+
+        // 2. Omitted
+        // 3.
+        float result = 0;
+        for (auto& flex_item : flex_items) {
+            auto product = 0;
+            if (flex_item.max_content_flex_fraction > 0) {
+                product = largest_max_content_flex_fraction * flex_item.box.computed_values().flex_grow_factor().value_or(1);
+            } else {
+                product = largest_max_content_flex_fraction * max(flex_item.box.computed_values().flex_shrink_factor().value_or(1), 1.0f) * flex_item.flex_base_size;
+            }
+            result += flex_item.flex_base_size + product;
+        }
+        main_available_size = clamp(result, main_min_size, main_max_size);
+    }
+    set_main_size(box, main_available_size);
+
+    /* 5. Collect flex items into flex lines: */
+    Vector<FlexLine> flex_lines {};
+    // FIXME: Also support wrap-reverse
+    if (box.computed_values().flex_wrap() == CSS::FlexWrap::Nowrap) {
+        FlexLine line = {};
+        for (auto& flex_item : flex_items) {
+            line.items.append(&flex_item);
```

Comment:
This doesn't seem very safe, the address of values in a `Vector<T>` isn't stable.
It doesn't _look_ like you're appending to `flex_items`, but I can't really tell for sure.

Reply:
The thing about this is: I sometimes need to iterate over _all_ items and sometimes they have to get iterated line-by-line. However both methods talk about the _same_ items. And both methods also may change attributes of the items.
I first gather all the items and put them into `flex_items` (Step 1). This `Vector<FlexItem>` never ever gets touched (ie appended to/deleted from) again.
I create the `FlexLine`s afterwards (Step 5) and they only get a pointer to the `FlexItem` inside `FlexItems`. Those Lines also don't get touched after they've been built.

I think this is a reasonable thing to do...

Reply:
Sure, it's alright to take references to the entries in a vector if you're not changing it.
One minor suggestion, then. it'd be nice if you could add a comment about not modifying the vector after some point (presumably the first time you try to take pointers to its entries).

Reply:
I'll do that! ;)
It would be nice to have a feature like `freeze()` or something to actively prohibit you from changing a Vector after it got initially filled. I don't know the overall usefulness of that outside this PR though

Reply:
Eh, I think you could do that by just wrapping the logic that creates that vector in a lambda and making it return a `Vector<T> const`.
Not really worth the pain though, a comment is enough most of the time

Reply:
`Not really worth the pain though` - I think so too.